### ISSUE TITLE
Recommended Security for X509 Certificates on IoT Edge Devices

### ISSUE THREAD

**Issue by srinathck** [Azure/iotedge#5166]
Using:
IoTEdge Version: 1.1
IoT Edge Modules Windows Containers on Windows Host Device running IoT Enterprise.

Summary:
I have Provisioned a Device using x509 Certificates using the Test Scripts mentioned [here](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-auto-provision-x509-certs?view=iotedge-2018-06&tabs=windows) and using this [doc](https://docs.microsoft.com/en-us/azure/iot-edge/iot-edge-certs?view=iotedge-2020-11)

We are going to use the Device Certificates to make a Mutual TLS calls to our External Api's.

I have a few questions as to how to handle these certificates.

1. Is it recommended that we save these cert and key (.pem) files on the host Device?
2. How can we make sure an attacker won't get ahold of these certificates and key?
3. Can these Device Certs be password protected? (Gave it a shot and Initialize-IoTEdge fails because the cert file is password protected. No way to pass in the password)
4. [Here ](https://docs.microsoft.com/en-us/azure/iot-edge/iot-edge-certs?view=iotedge-2020-11#device-ca-certificate)it talks about secure storage such as a hardware security module (HSM). I would like to learn more about how to setup a Soft HSM and not having to add/modify any actual hardware. Because honestly if I am going to have a Hardware module added, I rather just go for TPM provisioning. Due to cost reasons we would like to keep it within the device.

To summarize, How can I securely store Certs on a IoTEdge Device?

I might be missing something here and I appreciate any help with this. 
Thank you

Srinath

**Comment by srinathck** [comment]
@kgremban @varunpuranik @vipeller Can someone please take a look at this. Thanks

**Comment by eustacea** [comment]
Hi Srinath,

A good recommendation for IoTEdge Version 1.1 on Windows is to leverage the security features of the Windows Host OS for which IoTEdge becomes a protected resources within that environment, just like any other Windows app/software.  These may include enabling features like bitlocker and tightening access control policies.  Device certificates will reside in the filesystem but with the understanding that the filesystem is under protection of the security of the Windows OS.

A better recommendation in light of direct access to hardware HSM, if at all available for your scenario, is to upgrade to IoTEdge version 1.2 and beyond where we've built better extensions for secure access to the host subsystem including access to HSMs.  We achieve this through the [Azure IoT Identity Service](https://azure.github.io/iot-identity-service/) device security subsystem.  IoTEdge 1.2 is available now for Linux OS but if you must have Windows Host then watch for support for IoTEdge 1.2  on [Azure EFLOW](https://docs.microsoft.com/en-us/windows/iot/iot-enterprise/eflow).

Best regards,
Eustace

**Comment by srinathck** [comment]
Thanks for the Recommendations @eustacea. I will look into the suggestions and will do some research.
